### ISSUE TITLE
Enhancement: Restrict caret position

### ISSUE THREAD

**Issue by bfelder** [Wramberg/TerminalView#39]
In contrast to my system's shell and also in contrast to the rest of Sublime I can place the caret pretty much at any position.

Upon text input, the caret is then moved upwards either to right after the prompt or to the end of the last word if there is already input.

It would be better IMHO if the caret position would be restricted to right after the prompt, like in a normal terminal.

**Comment by Wramberg** [comment]
In order to support selection (to do copy/paste) the cursor position cannot be restricted. Other teminals implement this by allowing selection without moving the cursor. This is not possible within ST3 unfortunately (atleast not that I know of).

**Comment by bfelder** [comment]
OK. if restriction is not possible, maybe the caret could at least be automatically close in on the prompt.
Just check whether there is white space between current caret position and prompt/ end of last word.
What do think about this suggestion @Wramberg?

**Comment by Wramberg** [comment]
Im not sure I understand what you mean.

**Comment by Wramberg** [comment]
Note that the ST3 cursor is snapped back into position whenever you type or do anything in the terminal. Also - in the newest devel version you can hit `escape` to make the cursor jump back to its original position.

**Comment by bfelder** [comment]
$  <- prompt                      <- lots of white space ->                                       | <- caret, misplaced here by mouse click
$ | <- caret automatically moved here after mouse click, just like you just started to type or hit Esc

**Comment by bfelder** [comment]
It is a minor point though. Relevant only if you want to give TerrminalView an even more enhanced UX.

**Comment by Wramberg** [comment]
When you start typing the cursor moves back to the prompt ? if not then you are experiencing some kind of issue.

**Comment by bfelder** [comment]
Yes. Upon typing the cursor moves back to the prompt. No issue here. :+1: 
Still, I can place the caret far away from the prompt.
It would be best the cursor would move to the prompt right upon focusing of the TerminalView.
(See for example the behavior in a normal Sublime view. You may click somewhere yet the caret is moved to the very left - the first column.)

**Comment by Wramberg** [comment]
Oh I understand now :-) What we really want to do it get rid of the trailing spaces. The way the internal terminal emulator works is by using space as default char. So an empty screen is filled with spaces.

I've been doing a few tests with using a different default char and removing it when updating the terminal view. The only problem is which byte to use as "blank".

Lets leave this open and I'll give it a little more thought when I have time

**Comment by Wramberg** [comment]
Note that solving this issue would also fix the conflict reported in #18

**Comment by bfelder** [comment]
Great that I was finally able to express myself clearly. :-)
I have never seen those red lines. 

When you have solved the issue, just shoot me a message in case you need a Linux tester. 
Keep up the good work, knowing you have already created the best Terminal plugin for Sublime!

**Comment by Wramberg** [comment]
I have looked a bit more into this and actually have a working example. Problem is I have to do a lot patching and additional logic. Im hesitant in adding this cause it will increase maintenace overhead when upgrading pyte etc (I am overriding their default char amongst other things) . Im not sure if it would be worth it considering that it is a minor UX improvement - what do you think @bfelder?

**Comment by bfelder** [comment]
@Wramberg Thanks for letting me know. As far as I understand you are in the middle of upgrading to xterm/pyte (?? - I am not sure about which one). I do not understand enough about how this relates to the issue of caret positioning but the common sense of a developer dictates to set priorities wisely. In other words come back to the issue of caret positioning later after you have made the other more important changes. Now you have an understanding of the difficulty of implementing this UX feature. Until then keep this issue thread open.

**Comment by Wramberg** [comment]
Yes I am looking into removing pyte and using gateone terminal instead. The gateone terminal should give some wanted improvements like

 * xterm support
 * 256 color support
 * support for custom escape handlers

The problem is: With pyte it is "not so clean" to implement. With gateone it seems near impossible since it uses space in multiple places (none of which I can override from my code).

So I'll wait a bit until we (maybe) have gateone running and then I will have to decide whether to close this issue (without implementing it) or start patching gateone (tedious to maintain). Seeing as this is a minor improvement I'm leaning towards the former.

**Comment by bfelder** [comment]
@Wramberg This sounds like a sensible approach. I make you a deal: Take your time to properly implement gateone. Then see again how easy/difficult it actually is to implement the caret positioning. **If unsuccessful get in touch with me via this issue. Maybe I can pull a rabbit from my hat and find some elegant workaround.** Just keep this issue open until then as a channel of communication and reminder.

**Comment by f34nk** [comment]
Any news on this?
